File: src/lib/keys.ts
/**
 * Key names follow the legacy app (robotjs) so imported hotkeys keep working;
 * see src-tauri/src/input/keys.rs for the Rust side of the vocabulary.
 */

export const MODIFIERS = ["control", "alt", "shift", "command"] as const;
export type Modifier = (typeof MODIFIERS)[number];

const isMac = typeof navigator !== "undefined" && navigator.platform.toLowerCase().includes("mac");

const SPECIAL: Record<string, string> = {
  " ": "space",
  Enter: "enter",
  Tab: "tab",
  Escape: "escape",
  Backspace: "backspace",
  Delete: "delete",
  ArrowUp: "up",
  ArrowDown: "down",
  ArrowLeft: "left",
  ArrowRight: "right",
  Home: "home",
  End: "end",
  PageUp: "pageup",
  PageDown: "pagedown",
  Insert: "insert",
  CapsLock: "capslock",
  AudioVolumeMute: "audio_mute",
  AudioVolumeUp: "audio_vol_up",
  AudioVolumeDown: "audio_vol_down",
  MediaPlayPause: "audio_play",
  MediaTrackNext: "audio_next",
  MediaTrackPrevious: "audio_prev",
};

/** Main key of a keyboard event, or null for a bare modifier. */
export function keyFromEvent(e: KeyboardEvent): string | null {
  if (["Control", "Alt", "Shift", "Meta", "AltGraph"].includes(e.key)) return null;
  if (e.code.startsWith("Numpad")) {
    const rest = e.code.slice(6);
    if (/^\d$/.test(rest)) return `numpad_${rest}`;
    const map: Record<string, string> = { Add: "numpad_+", Subtract: "numpad_-", Multiply: "numpad_*", Divide: "numpad_/", Decimal: "numpad_." };
    if (map[rest]) return map[rest];
  }
  if (SPECIAL[e.key]) return SPECIAL[e.key];
  if (/^F\d{1,2}$/.test(e.key)) return e.key.toLowerCase();
  if (e.key.length === 1) return e.key.toLowerCase();
  return null;
}

const MOUSE_BUTTONS: Record<number, string> = { 1: "mouse_middle", 2: "mouse_right", 3: "mouse_back", 4: "mouse_forward" };

/** A mouse button usable like a key (side buttons for push-to-talk); the left button is never captured. */
export function mouseFromEvent(e: MouseEvent): string | null {
  return MOUSE_BUTTONS[e.button] ?? null;
}

export function isMouseKey(key: string): boolean {
  return key.startsWith("mouse_");
}

export function modifiersFromEvent(e: { ctrlKey: boolean; altKey: boolean; shiftKey: boolean; metaKey: boolean }): Modifier[] {
  const mods: Modifier[] = [];
  if (e.ctrlKey) mods.push("control");
  if (e.altKey) mods.push("alt");
  if (e.shiftKey) mods.push("shift");
  if (e.metaKey) mods.push("command");
  return mods;
}

export function formatModifier(m: string): string {
  switch (m) {
    case "control":
      return isMac ? "⌃" : "Ctrl";
    case "alt":
      return isMac ? "⌥" : "Alt";
    case "shift":
      return isMac ? "⇧" : "Shift";
    case "command":
      return isMac ? "⌘" : "Win";
    default:
      return m;
  }
}

export function formatKey(key: string): string {
  if (!key) return "";
  if (key.length === 1) return key.toUpperCase();
  if (/^f\d{1,2}$/.test(key)) return key.toUpperCase();
  const names: Record<string, string> = {
    space: "Space",
    enter: "Enter",
    tab: "Tab",
    escape: "Esc",
    backspace: "Backspace",
    delete: "Delete",
    up: "↑",
    down: "↓",
    left: "←",
    right: "→",
    home: "Home",
    end: "End",
    pageup: "Page Up",
    pagedown: "Page Down",
    insert: "Insert",
    capslock: "Caps Lock",
  };
  if (names[key]) return names[key];
  if (key.startsWith("mouse_")) {
    const which: Record<string, string> = { left: "left", right: "right", middle: "middle", back: "back", forward: "forward", "4": "back", "5": "forward" };
    return `Mouse ${which[key.slice(6)] ?? key.slice(6)}`;
  }
  if (key.startsWith("numpad_")) return `Num ${key.slice(7)}`;
  const media: Record<string, string> = { audio_mute: "Mute", audio_vol_down: "Volume down", audio_vol_up: "Volume up", audio_play: "Play/Pause", audio_next: "Next track", audio_prev: "Previous track" };
  if (media[key]) return media[key];
  if (key.startsWith("audio_")) return key.slice(6).replace("_", " ");
  return key;
}

export function formatCombo(key: string, modifiers: string[]): string {
  const parts = modifiers.map(formatModifier);
  if (key) parts.push(formatKey(key));
  return parts.join(isMac ? "" : " + ");
}

export type KeyGroupId = "function" | "navigation" | "editing" | "numpad" | "media" | "mouse" | "letters" | "digits";

export interface KeyGroup {
  id: KeyGroupId;
  keys: string[];
}

/** Windows presses F1 to F24; macOS's key codes stop at F20 (see function_key in keys.rs). */
const FUNCTION_KEYS = isMac ? 20 : 24;

/**
 * Every key name the app can press, grouped for the picker: the way to set a key
 * the keyboard cannot produce (F13 on a laptop, Escape, a missing number pad).
 */
export const KEY_GROUPS: KeyGroup[] = [
  { id: "function", keys: Array.from({ length: FUNCTION_KEYS }, (_, i) => `f${i + 1}`) },
  { id: "navigation", keys: ["up", "down", "left", "right", "home", "end", "pageup", "pagedown"] },
  { id: "editing", keys: ["space", "enter", "tab", "escape", "backspace", "delete", "insert", "capslock"] },
  { id: "numpad", keys: [...Array.from({ length: 10 }, (_, i) => `numpad_${i}`), "numpad_+", "numpad_-", "numpad_*", "numpad_/", "numpad_."] },
  { id: "media", keys: ["audio_mute", "audio_vol_down", "audio_vol_up", "audio_play", "audio_next", "audio_prev"] },
  { id: "mouse", keys: ["mouse_left", "mouse_middle", "mouse_right", "mouse_back", "mouse_forward"] },
  { id: "letters", keys: "abcdefghijklmnopqrstuvwxyz".split("") },
  { id: "digits", keys: "0123456789".split("") },
];

File: src-tauri/src/input/keys.rs
//! Key vocabulary. Names follow the legacy app (robotjs) so imported hotkeys
//! keep working: letters and digits, `f1`..`f20`, `space`, `enter`, `tab`,
//! `escape`, `backspace`, `delete`, `up`/`down`/`left`/`right`, `home`, `end`,
//! `pageup`, `pagedown`, `capslock`, `numpad_0`..`numpad_9`, `numpad_+` etc.,
//! `audio_*`, and the modifiers `control`, `alt`, `shift`, `command`.

use enigo::{Button, Key};

/// Mouse buttons usable wherever a key is (push-to-talk, hotkey steps).
pub fn parse_mouse(name: &str) -> Option<Button> {
    match name.trim().to_lowercase().as_str() {
        "mouse_left" => Some(Button::Left),
        "mouse_right" => Some(Button::Right),
        "mouse_middle" => Some(Button::Middle),
        "mouse_back" | "mouse_4" => Some(Button::Back),
        "mouse_forward" | "mouse_5" => Some(Button::Forward),
        _ => None,
    }
}

pub fn parse_key(name: &str) -> Option<Key> {
    let n = name.trim().to_lowercase();
    if let Some(m) = parse_modifier(&n) {
        return Some(m);
    }
    let key = match n.as_str() {
        "space" => Key::Space,
        "enter" | "return" => Key::Return,
        "tab" => Key::Tab,
        "escape" | "esc" => Key::Escape,
        "backspace" => Key::Backspace,
        "delete" | "del" => Key::Delete,
        "up" => Key::UpArrow,
        "down" => Key::DownArrow,
        "left" => Key::LeftArrow,
        "right" => Key::RightArrow,
        "home" => Key::Home,
        "end" => Key::End,
        "pageup" => Key::PageUp,
        "pagedown" => Key::PageDown,
        "capslock" => Key::CapsLock,
        "help" | "insert" => Key::Help,
        "audio_mute" => Key::VolumeMute,
        "audio_vol_down" => Key::VolumeDown,
        "audio_vol_up" => Key::VolumeUp,
        "audio_play" | "audio_pause" => Key::MediaPlayPause,
        "audio_next" => Key::MediaNextTrack,
        "audio_prev" => Key::MediaPrevTrack,
        "numpad_0" => Key::Numpad0,
        "numpad_1" => Key::Numpad1,
        "numpad_2" => Key::Numpad2,
        "numpad_3" => Key::Numpad3,
        "numpad_4" => Key::Numpad4,
        "numpad_5" => Key::Numpad5,
        "numpad_6" => Key::Numpad6,
        "numpad_7" => Key::Numpad7,
        "numpad_8" => Key::Numpad8,
        "numpad_9" => Key::Numpad9,
        "numpad_+" => Key::Add,
        "numpad_-" => Key::Subtract,
        "numpad_*" => Key::Multiply,
        "numpad_/" => Key::Divide,
        "numpad_." => Key::Decimal,
        _ => {
            if let Some(num) = n.strip_prefix('f').and_then(|d| d.parse::<u8>().ok()) {
                return function_key(num);
            }
            let mut chars = name.trim().chars();
            match (chars.next(), chars.next()) {
                (Some(c), None) => Key::Unicode(c.to_ascii_lowercase()),
                _ => return None,
            }
        }
    };
    Some(key)
}

pub fn parse_modifier(name: &str) -> Option<Key> {
    Some(match name.trim().to_lowercase().as_str() {
        "control" | "ctrl" | "left_control" => Key::Control,
        "right_control" => Key::RControl,
        "alt" | "option" => Key::Alt,
        "right_alt" => Key::Alt,
        "shift" | "left_shift" => Key::Shift,
        "right_shift" => Key::RShift,
        "command" | "cmd" | "meta" | "super" | "win" | "windows" => Key::Meta,
        _ => return None,
    })
}

fn function_key(n: u8) -> Option<Key> {
    Some(match n {
        1 => Key::F1,
        2 => Key::F2,
        3 => Key::F3,
        4 => Key::F4,
        5 => Key::F5,
        6 => Key::F6,
        7 => Key::F7,
        8 => Key::F8,
        9 => Key::F9,
        10 => Key::F10,
        11 => Key::F11,
        12 => Key::F12,
        13 => Key::F13,
        14 => Key::F14,
        15 => Key::F15,
        16 => Key::F16,
        17 => Key::F17,
        18 => Key::F18,
        19 => Key::F19,
        20 => Key::F20,
        _ => return None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_common_names() {
        assert_eq!(parse_key("f9"), Some(Key::F9));
        assert_eq!(parse_key("a"), Some(Key::Unicode('a')));
        assert_eq!(parse_key("A"), Some(Key::Unicode('a')));
        assert_eq!(parse_key("space"), Some(Key::Space));
        assert_eq!(parse_key("numpad_+"), Some(Key::Add));
        assert_eq!(parse_modifier("command"), Some(Key::Meta));
        assert_eq!(parse_key("f99"), None);
        assert_eq!(parse_key("nonsense"), None);
    }
}
